File: unity/src/project/SteamAudioUnity/Assets/Plugins/SteamAudio/Scripts/Runtime/SteamAudioSettings.cs
//
// Copyright 2017-2023 Valve Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

using UnityEngine;
#if UNITY_EDITOR
using UnityEditor;
#endif

namespace SteamAudio
{
    public enum AudioEngineType
    {
        Unity,
        FMODStudio,
        Wwise
    }

    [CreateAssetMenu(menuName = "Steam Audio/Steam Audio Settings")]
    public class SteamAudioSettings : ScriptableObject
    {
        [Header("Audio Engine Settings")]
        public AudioEngineType audioEngine = AudioEngineType.Unity;

        [Header("HRTF Settings")]
        public bool hrtfDisabled = false;
        public bool perspectiveCorrection = false;
        [Range(.25f, 4.0f)]
        public float perspectiveCorrectionFactor = 1.0f;
        [Range(-12.0f, 12.0f)]
        public float hrtfVolumeGainDB = 0.0f;
        public HRTFNormType hrtfNormalizationType = HRTFNormType.None;
        public SOFAFile[] SOFAFiles = null;

        [Header("Material Settings")]
        public SteamAudioMaterial defaultMaterial = null;

        [Header("Ray Tracer Settings")]
        public SceneType sceneType = SceneType.Default;
        public LayerMask layerMask = new LayerMask();

        [Header("Occlusion Settings")]
        [Range(1, 128)]
        public int maxOcclusionSamples = 16;

        [Header("Real-time Reflections Settings")]
        [Range(1024, 65536)]
        public int realTimeRays = 4096;
        [Range(1, 64)]
        public int realTimeBounces = 4;
        [Range(0.1f, 10.0f)]
        public float realTimeDuration = 1.0f;
        [Range(0, 3)]
        public int realTimeAmbisonicOrder = 1;
        [Range(1, 128)]
        public int realTimeMaxSources = 32;
        [Range(0, 100)]
        public int realTimeCPUCoresPercentage = 5;
        [Range(0.1f, 10.0f)]
        public float realTimeIrradianceMinDistance = 1.0f;

        [Header("Baked Reflections Settings")]
        public bool bakeConvolution = true;
        public bool bakeParametric = false;
        [Range(1024, 65536)]
        public int bakingRays = 16384;
        [Range(1, 64)]
        public int bakingBounces = 16;
        [Range(0.1f, 10.0f)]
        public float bakingDuration = 1.0f;
        [Range(0, 3)]
        public int bakingAmbisonicOrder = 1;
        [Range(0, 100)]
        public int bakingCPUCoresPercentage = 50;
        [Range(0.1f, 10.0f)]
        public float bakingIrradianceMinDistance = 1.0f;

        [Header("Baked Pathing Settings")]
        [Range(1, 32)]
        public int bakingVisibilitySamples = 4;
        [Range(0.0f, 2.0f)]
        public float bakingVisibilityRadius = 1.0f;
        [Range(0.0f, 1.0f)]
        public float bakingVisibilityThreshold = 0.1f;
        [Range(0.0f, 1000.0f)]
        public float bakingVisibilityRange = 1000.0f;
        [Range(0.0f, 1000.0f)]
        public float bakingPathRange = 1000.0f;
        [Range(0, 100)]
        public int bakedPathingCPUCoresPercentage = 50;

        [Header("Simulation Update Settings")]
        [Range(0.1f, 1.0f)]
        public float simulationUpdateInterval = 0.1f;

        [Header("Reflection Effect Settings")]
        public ReflectionEffectType reflectionEffectType = ReflectionEffectType.Convolution;

        [Header("Hybrid Reverb Settings")]
        [Range(0.1f, 2.0f)]
        public float hybridReverbTransitionTime = 1.0f;
        [Range(0, 100)]
        public int hybridReverbOverlapPercent = 25;

        [Header("OpenCL Settings")]
        public OpenCLDeviceType deviceType = OpenCLDeviceType.GPU;
        [Range(0, 16)]
        public int maxReservedComputeUnits = 8;
        [Range(0.0f, 1.0f)]
        public float fractionComputeUnitsForIRUpdate = 0.5f;

        [Header("Radeon Rays Settings")]
        [Range(1, 16)]
        public int bakingBatchSize = 8;

        [Header("TrueAudio Next Settings")]
        [Range(0.1f, 10.0f)]
        public float TANDuration = 1.0f;
        [Range(0, 3)]
        public int TANAmbisonicOrder = 1;
        [Range(1, 128)]
        public int TANMaxSources = 32;

        [Header("Advanced Settings")]
        public bool EnableValidation = false;

        static SteamAudioSettings sSingleton = null;

        public static SteamAudioSettings Singleton
        {
            get
            {
                if (sSingleton == null)
                {
                    sSingleton = Resources.Load<SteamAudioSettings>("SteamAudioSettings");
                    if (sSingleton == null)
                    {
                        sSingleton = CreateInstance<SteamAudioSettings>();
                        sSingleton.name = "Steam Audio Settings";

#if UNITY_EDITOR
                        sSingleton.defaultMaterial = (SteamAudioMaterial) AssetDatabase.LoadAssetAtPath("Assets/Plugins/SteamAudio/Resources/Materials/Default.asset", typeof(SteamAudioMaterial));

                        AssetDatabase.CreateAsset(sSingleton, "Assets/Plugins/SteamAudio/Resources/SteamAudioSettings.asset");
#endif
                    }
                }

                return sSingleton;
            }
        }
    }
}

File: unreal/src/SteamAudioUnreal/Plugins/SteamAudio/Source/SteamAudio/Private/SteamAudioSettings.cpp
//
// Copyright 2017-2023 Valve Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#include "SteamAudioSettings.h"
#include "SteamAudioMaterial.h"
#include "SOFAFile.h"

// ---------------------------------------------------------------------------------------------------------------------
// USteamAudioSettings
// ---------------------------------------------------------------------------------------------------------------------

USteamAudioSettings::USteamAudioSettings()
    : AudioEngine(EAudioEngineType::UNREAL)
    , bExportLandscapeGeometry(true)
    , bExportBSPGeometry(true)
    , DefaultMeshMaterial("/SteamAudio/Materials/Default.Default")
    , DefaultLandscapeMaterial("/SteamAudio/Materials/Default.Default")
    , DefaultBSPMaterial("/SteamAudio/Materials/Default.Default")
    , SceneType(ESceneType::DEFAULT)
    , MaxOcclusionSamples(16)
    , RealTimeRays(4096)
    , RealTimeBounces(4)
    , RealTimeDuration(1.0f)
    , RealTimeAmbisonicOrder(1)
    , RealTimeMaxSources(32)
    , RealTimeCPUCoresPercentage(5)
    , RealTimeIrradianceMinDistance(1.0f)
    , bBakeConvolution(true)
    , bBakeParametric(false)
    , BakingRays(16384)
    , BakingBounces(16)
    , BakingDuration(1.0f)
    , BakingAmbisonicOrder(1)
    , BakingCPUCoresPercentage(50)
    , BakingIrradianceMinDistance(1.0f)
    , ReverbSubmix(nullptr)
    , BakingVisibilitySamples(4)
    , BakingVisibilityRadius(1.0f)
    , BakingVisibilityThreshold(0.1f)
    , BakingVisibilityRange(1000.0f)
    , BakingPathRange(1000.0f)
    , BakedPathingCPUCoresPercentage(50)
    , SimulationUpdateInterval(0.1f)
    , ReflectionEffectType(EReflectionEffectType::CONVOLUTION)
    , HybridReverbTransitionTime(1.0f)
    , HybridReverbOverlapPercent(25)
    , DeviceType(EOpenCLDeviceType::ANY)
    , MaxReservedComputeUnits(8)
    , FractionComputeUnitsForIRUpdate(0.5f)
    , BakingBatchSize(8)
    , TANDuration(1.0f)
    , TANAmbisonicOrder(1)
    , TANMaxSources(32)
    , HRTFVolume(0.0f)
    , HRTFNormalizationType(EHRTFNormType::NONE)
    , SOFAFile(nullptr)
    , EnableValidation(false)
{}

FSteamAudioSettings USteamAudioSettings::GetSettings() const
{
    FSteamAudioSettings Settings{};
    Settings.AudioEngine = AudioEngine;
    Settings.bExportLandscapeGeometry = bExportLandscapeGeometry;
    Settings.bExportBSPGeometry = bExportBSPGeometry;
    Settings.DefaultMeshMaterial = GetMaterialForAsset(DefaultMeshMaterial);
    Settings.DefaultLandscapeMaterial = GetMaterialForAsset(DefaultLandscapeMaterial);
    Settings.DefaultBSPMaterial = GetMaterialForAsset(DefaultBSPMaterial);
    Settings.SceneType = static_cast<IPLSceneType>(SceneType);
    Settings.MaxOcclusionSamples = MaxOcclusionSamples;
    Settings.RealTimeRays = RealTimeRays;
    Settings.RealTimeBounces = RealTimeBounces;
    Settings.RealTimeDuration = RealTimeDuration;
    Settings.RealTimeAmbisonicOrder = RealTimeAmbisonicOrder;
    Settings.RealTimeMaxSources = RealTimeMaxSources;
    Settings.RealTimeCPUCoresPercentage = RealTimeCPUCoresPercentage;
    Settings.RealTimeIrradianceMinDistance = RealTimeIrradianceMinDistance;
    Settings.bBakeConvolution = bBakeConvolution;
    Settings.bBakeParametric = bBakeParametric;
    Settings.BakingRays = BakingRays;
    Settings.BakingBounces = BakingBounces;
    Settings.BakingDuration = BakingDuration;
    Settings.BakingAmbisonicOrder = BakingAmbisonicOrder;
    Settings.BakingCPUCoresPercentage = BakingCPUCoresPercentage;
    Settings.BakedPathingCPUCoresPercentage = BakedPathingCPUCoresPercentage;
    Settings.ReverbSubmix = nullptr;
    Settings.BakingVisibilitySamples = BakingVisibilitySamples;
    Settings.BakingVisibilityRadius = BakingVisibilityRadius;
    Settings.BakingVisibilityThreshold = BakingVisibilityThreshold;
    Settings.BakingVisibilityRange = BakingVisibilityRange;
    Settings.BakingPathRange = BakingPathRange;
    Settings.BakedPathingCPUCoresPercentage = BakedPathingCPUCoresPercentage;
    Settings.SimulationUpdateInterval = SimulationUpdateInterval;
    Settings.ReflectionEffectType = static_cast<IPLReflectionEffectType>(ReflectionEffectType);
    Settings.HybridReverbTransitionTime = HybridReverbTransitionTime;
    Settings.HybridReverbOverlapPercent = HybridReverbOverlapPercent;
    Settings.OpenCLDeviceType = static_cast<IPLOpenCLDeviceType>(DeviceType);
    Settings.MaxReservedComputeUnits = MaxReservedComputeUnits;
    Settings.FractionComputeUnitsForIRUpdate = FractionComputeUnitsForIRUpdate;
    Settings.BakingBatchSize = BakingBatchSize;
    Settings.TANDuration = TANDuration;
    Settings.TANAmbisonicOrder = TANAmbisonicOrder;
    Settings.TANMaxSources = TANMaxSources;
    Settings.SOFAFile = Cast<USOFAFile>(SOFAFile.TryLoad());
    Settings.HRTFVolume = (Settings.SOFAFile) ? Settings.SOFAFile->Volume : HRTFVolume;
    Settings.HRTFNormType = (Settings.SOFAFile) ? static_cast<IPLHRTFNormType>(Settings.SOFAFile->NormalizationType) : static_cast<IPLHRTFNormType>(HRTFNormalizationType);
    return Settings;
}

IPLMaterial USteamAudioSettings::GetMaterialForAsset(FSoftObjectPath Asset) const
{
    IPLMaterial SteamAudioMaterial{};
    SteamAudioMaterial.absorption[0] = 0.1f;
    SteamAudioMaterial.absorption[1] = 0.1f;
    SteamAudioMaterial.absorption[2] = 0.1f;
    SteamAudioMaterial.scattering = 0.5f;
    SteamAudioMaterial.transmission[0] = 0.1f;
    SteamAudioMaterial.transmission[1] = 0.1f;
    SteamAudioMaterial.transmission[2] = 0.1f;

    USteamAudioMaterial* Material = Cast<USteamAudioMaterial>(Asset.TryLoad());
    if (Material)
    {
        SteamAudioMaterial.absorption[0] = Material->AbsorptionLow;
        SteamAudioMaterial.absorption[1] = Material->AbsorptionMid;
        SteamAudioMaterial.absorption[2] = Material->AbsorptionHigh;
        SteamAudioMaterial.scattering = Material->Scattering;
        SteamAudioMaterial.transmission[0] = Material->TransmissionLow;
        SteamAudioMaterial.transmission[1] = Material->TransmissionMid;
        SteamAudioMaterial.transmission[2] = Material->TransmissionHigh;
    }

    return SteamAudioMaterial;
}

UObject* USteamAudioSettings::GetObjectForAsset(FSoftObjectPath Asset) const
{
    return Asset.TryLoad();
}
